### PR TITLE
Implement P2443R1 views::chunk_by

### LINKED ISSUE #2540: P2443R1 `views::chunk_by`
[P2443R1](https://wg21.link/P2443R1) `views::chunk_by`

Feature-test macro:
```cpp
#define __cpp_lib_ranges_chunk_by 202202L // need to confirm official value
```

### PR DESCRIPTION
**Pull request by cpplearner**
This implements [P2443R1 `views::chunk_by`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2443r1.html).

Fixes #2540.

### CODE REVIEW

File: stl/inc/ranges [hunk 1]
```diff
@@ -4741,6 +4741,187 @@ namespace ranges {
...
+            chunk_by_view* _Parent{};
+            /* [[no_unique_address]] */ iterator_t<_Vw> _Current{};
+            /* [[no_unique_address]] */ iterator_t<_Vw> _Next{};
+
+            constexpr _Iterator(chunk_by_view& _Parent_, iterator_t<_Vw> _Current_, iterator_t<_Vw> _Next_) noexcept(
+                is_nothrow_move_constructible_v<iterator_t<_Vw>>) // strengthened
+                : _Parent(_STD addressof(_Parent_)), _Current(_STD move(_Current_)), _Next(_STD move(_Next_)) {}
+
+        public:
+            using value_type        = subrange<iterator_t<_Vw>>;
```
Comment (miscco) [inline stl/inc/ranges:4766]: I am really sad that we do not have reconstructible ranges :(

File: stl/inc/ranges [hunk 2]
```diff
@@ -4741,6 +4741,187 @@ namespace ranges {
...
+            using difference_type   = range_difference_t<_Vw>;
+            using iterator_category = input_iterator_tag;
+            using iterator_concept =
+                conditional_t<bidirectional_range<_Vw>, bidirectional_iterator_tag, forward_iterator_tag>;
+
+            _Iterator() = default;
+
+            _NODISCARD constexpr value_type operator*() const {
```
Comment (miscco) [inline stl/inc/ranges:4774]: I wanted to complain about lack of `noexcept(is_nothrow_copy_constructible_v<iterator_t<_Vw>>) /* strengthen */`

That said it seems we were a bit lazy there with `subrange` Should we still add it given that we pretty sure know whether it may throw or not?

File: stl/inc/ranges [hunk 3]
```diff
@@ -4741,6 +4741,187 @@ namespace ranges {
...
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+                return subrange(_Current, _Next);
+            }
+
+            constexpr _Iterator& operator++() {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                _STL_VERIFY(_Current != _Next, "cannot increment chunk_by_view iterator past end");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+                _Current = _Next;
+                _Next    = _Parent->_Find_next(_Current);
```
Comment (miscco) [inline stl/inc/ranges:4786]: This is correct, but gave me a little pause, because I had expected `_Next` as the argument not the updated `_Current`.

File: stl/inc/ranges [hunk 4]
```diff
@@ -4741,6 +4741,187 @@ namespace ranges {
...
+
+            constexpr _Iterator operator++(int) {
+                auto _Tmp = *this;
+                ++*this;
+                return _Tmp;
+            }
+
+            constexpr _Iterator& operator--() requires bidirectional_range<_Vw> {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                _STL_VERIFY(_Parent != nullptr, "cannot decrement value-initialized chunk_by_view iterator");
```
Comment (miscco) [inline stl/inc/ranges:4798]: Should we add a check that `_Current` is not equal to `_Parent->begin()`?

EDIT: I see that you put it in `_Find_prev` Given that there is only one occurrence of that I would prefer to pull it up here

File: stl/inc/ranges [hunk 5]
```diff
@@ -4741,6 +4741,187 @@ namespace ranges {
...
+                return _Left._Current == _Left._Next;
+            }
+        };
+
+        _NODISCARD constexpr iterator_t<_Vw> _Find_next(iterator_t<_Vw> _It) {
+#if _ITERATOR_DEBUG_LEVEL != 0
+            _STL_VERIFY(_Pred, "cannot increment a chunk_by_view iterator whose parent view has no predicate");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+
+            return _RANGES next(_RANGES adjacent_find(_It, _RANGES end(_Range),
```
Comment (miscco) [inline stl/inc/ranges:4827]: Can we pull that out into multiple statements, it is really hard to follow that

File: stl/inc/ranges [hunk 6]
```diff
@@ -4741,6 +4741,187 @@ namespace ranges {
...
+                                    [this]<class _Ty1, class _Ty2>(_Ty1&& _Left, _Ty2&& _Right) {
```
Comment (miscco) [inline stl/inc/ranges:4828]: We could only capture `*_Pred` by value to reduce scope

File: stl/inc/ranges [hunk 7]
```diff
@@ -4741,6 +4741,187 @@ namespace ranges {
...
+                                        return !_STD invoke(
+                                            *_Pred, _STD forward<_Ty2>(_Right), _STD forward<_Ty1>(_Left));
+                                    })
+                                    .base(),
+                1, _RANGES begin(_Range));
+        }
+
+    public:
+        // clang-format off
+        chunk_by_view() requires default_initializable<_Vw> && default_initializable<_Vw> = default;
```
Comment (miscco) [inline stl/inc/ranges:4853]: That is a lot of `default_initializable<_Vw>`
```suggestion
        chunk_by_view() requires default_initializable<_Vw> && default_initializable<_Pr> = default;
```

**Review by miscco — COMMENTED** [review]
I wanted to start this yesterday evening but my wife saved me with watching "The expanse", so I am double happy

File: stl/inc/ranges (hunk 3 above)
Comment (cpplearner) [reply stl/inc/ranges:4786]: This is what P2443R1 says, but I also prefer using `_Next`.

File: stl/inc/ranges (hunk 6 above)
Comment (cpplearner) [reply stl/inc/ranges:4828]: I don't think we can capture `*_Pred` by copy since it might not be copyable, but by reference should be find.

File: stl/inc/ranges [hunk 8]
```diff
@@ -4741,6 +4741,186 @@ namespace ranges {
...
+            const auto _Before_next = _RANGES adjacent_find(_It, _RANGES end(_Range), _Not_pred);
+            return _RANGES next(_Before_next, 1, _RANGES end(_Range));
+        }
+
+        _NODISCARD constexpr iterator_t<_Vw> _Find_prev(iterator_t<_Vw> _It) {
+#if _ITERATOR_DEBUG_LEVEL != 0
+            _STL_VERIFY(_Pred, "cannot decrement a chunk_by_view iterator whose parent view has no predicate");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+
+            reverse_view _Rv{subrange{_RANGES begin(_Range), _It}};
```
Comment (StephanTLavavej) [inline stl/inc/ranges:4842]: I observe that this constructs `subrange` with braces, but earlier a `subrange` was constructed with parens. They should be consistent (as mentioned in my other comment, I don't know which form is best).

File: stl/inc/ranges [hunk 9]
```diff
@@ -4741,6 +4741,186 @@ namespace ranges {
...
+#if _CONTAINER_DEBUG_LEVEL > 0
+            _STL_VERIFY(_Pred, "cannot call begin on a chunk_by_view with no predicate");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+            if (this->_Has_cache()) {
+                return _Iterator{*this, _RANGES begin(_Range), this->_Get_cache(_Range)};
+            }
+
+            auto _Result = _Find_next(_RANGES begin(_Range));
+            this->_Set_cache(_Range, _Result);
+            return _Iterator{*this, _RANGES begin(_Range), _Result};
```
Comment (StephanTLavavej) [inline stl/inc/ranges:4884]: This function mentions `_RANGES begin(_Range)` 3 times, and always executes 1 or 2 of them. Would lifting this out be a good idea?

File: stl/inc/ranges [hunk 10]
```diff
@@ -4741,6 +4741,186 @@ namespace ranges {
...
+        }
+
+        _NODISCARD constexpr auto end() {
+            if constexpr (common_range<_Vw>) {
+                return _Iterator{*this, _RANGES end(_Range), _RANGES end(_Range)};
```
Comment (StephanTLavavej) [inline stl/inc/ranges:4889]: Would extracting `_RANGES end(_Range)` be helpful?

File: stl/inc/ranges [hunk 11]
```diff
@@ -4741,6 +4741,186 @@ namespace ranges {
...
+    namespace views {
+        struct _Chunk_by_fn {
+            // clang-format off
+            template <viewable_range _Rng, class _Pr>
+            _NODISCARD constexpr auto operator()(_Rng&& _Range, _Pr&& _Pred) const noexcept(
+                noexcept(chunk_by_view(_STD forward<_Rng>(_Range), _STD forward<_Pr>(_Pred)))) requires requires {
+                chunk_by_view(_STD forward<_Rng>(_Range), _STD forward<_Pr>(_Pred));
+            } {
+                // clang-format on
+                return chunk_by_view(_STD forward<_Rng>(_Range), _STD forward<_Pr>(_Pred));
```
Comment (StephanTLavavej) [inline stl/inc/ranges:4908]: I don't understand why, but I have observed that other `views` will switch to saying `static_cast<T&&>(t)` in their `requires`, despite our usual currently preferred convention of using `forward` elsewhere. Not sure if this matters here. Example:
https://github.com/microsoft/STL/blob/afe0800cd6bf0d163bf678ad4e1feb3c130d1ec5/stl/inc/ranges#L3806-L3815

File: tests/std/tests/P2443R1_views_chunk_by_death/test.cpp [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    (void) r.pred(); // chunk_by_view has no predicate
+}
+
+void test_view_begin() {
+    auto r = with_no_predicate();
+    (void) r.begin(); // cannot call begin on a chunk_by_view that holds no predicate
+}
+
+constexpr auto lambda = [](int i, int j) { return i == j; };
+using FV              = decltype(ranges::chunk_by_view{some_ints, lambda});
```
Comment (StephanTLavavej) [inline tests/std/tests/P2443R1_views_chunk_by_death/test.cpp:51]: This was copy-pasted from:
https://github.com/microsoft/STL/blob/afe0800cd6bf0d163bf678ad4e1feb3c130d1ec5/tests/std/tests/P0896R4_views_filter_death/test.cpp#L51
The abbreviation `FV` should be changed.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/ranges [hunk 13]
```diff
@@ -4741,6 +4741,186 @@ namespace ranges {
...
+
+            constexpr _Iterator operator++(int) {
+                auto _Tmp = *this;
+                ++*this;
+                return _Tmp;
+            }
+
+            constexpr _Iterator& operator--() requires bidirectional_range<_Vw> {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                _STL_VERIFY(_Parent != nullptr, "cannot decrement value-initialized chunk_by_view iterator");
```
[the remark this answers is not in the document]
Comment (cpplearner) [reply stl/inc/ranges:4799]: I believe that a value-initialized chunk_by_view iterator is always an end iterator (because its `_Current` and `_Next` are equal, because both are value-initialized), so `operator++` only needs to check `_Current != _Next`.

I'd prefer to keep the null check in `operator--`, otherwise it would seem that the next line is introducing an error (null dereference) when checking for another error.

File: stl/inc/ranges (hunk 8 above)
Comment (cpplearner) [reply stl/inc/ranges:4842]: I'll choose braces.

File: stl/inc/ranges (hunk 11 above)
Comment (cpplearner) [reply stl/inc/ranges:4908]: And `filter` has mixed `static_cast` and `_STD forward`.
https://github.com/microsoft/STL/blob/afe0800cd6bf0d163bf678ad4e1feb3c130d1ec5/stl/inc/ranges#L1841-L1844

And `transform` even uses `_STD move`.
https://github.com/microsoft/STL/blob/afe0800cd6bf0d163bf678ad4e1feb3c130d1ec5/stl/inc/ranges#L2300-L2303

@CaseyCarter do you remember why this was done in this way?

File: stl/inc/ranges [hunk 14]
```diff
@@ -4741,6 +4741,186 @@ namespace ranges {
...
         inline constexpr auto values = elements<1>;
     } // namespace views
+
+#if _HAS_CXX23
+    template <forward_range _Vw, indirect_binary_predicate<iterator_t<_Vw>, iterator_t<_Vw>> _Pr>
+        requires view<_Vw> && is_object_v<_Pr>
+    class chunk_by_view : public _Cached_position<_Vw, chunk_by_view<_Vw, _Pr>> {
+    private:
+        /* [[no_unique_address]] */ _Vw _Range{};
+        /* [[no_unique_address]] */ _Copyable_box<_Pr> _Pred{};
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/ranges:4751]: Sounds great, thanks for checking the existing convention!

File: stl/inc/ranges (hunk 13 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/ranges:4799]: Ah, that makes sense. Sounds good to me.

**Review by StephanTLavavej — APPROVED** [review]

File: tests/std/tests/P2443R1_views_chunk_by_death/test.cpp [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+    auto r = with_no_predicate();
+    (void) r.pred(); // chunk_by_view has no predicate
+}
+
+void test_view_begin() {
+    auto r = with_no_predicate();
+    (void) r.begin(); // cannot call begin on a chunk_by_view that holds no predicate
+}
+
+constexpr auto lambda = [](int i, int j) { return i == j; };
```
Comment (CaseyCarter) [inline tests/std/tests/P2443R1_views_chunk_by_death/test.cpp:50]: Would it be simpler to spell this `std::equal_to<int>{}`? (Pure style comment, feel free to ignore if you like.)

File: stl/inc/ranges [hunk 16]
```diff
@@ -4741,6 +4741,189 @@ namespace ranges {
...
+            }
+
+            constexpr _Iterator operator--(int) requires bidirectional_range<_Vw> {
+                auto _Tmp = *this;
+                --*this;
+                return _Tmp;
+            }
+
+            _NODISCARD friend constexpr bool operator==(const _Iterator& _Left, const _Iterator& _Right) noexcept(
+                noexcept(_Left._Current == _Right._Current)) /* strengthened */ {
```
[the remark this answers is not in the document]
Comment (cpplearner) [reply stl/inc/ranges:4815]: I'll follow https://github.com/microsoft/STL/blob/dafa1dd5fae2bfc064b64449a63f386b0e148b08/stl/inc/ranges#L3286-L3288 to use `_Implicitly_convert_to`.

File: tests/std/tests/P2443R1_views_chunk_by_death/test.cpp (hunk 15 above)
Comment (cpplearner) [reply tests/std/tests/P2443R1_views_chunk_by_death/test.cpp]: The behavior of this predicate doesn't matter. I'll change it to `return true;` to make it look more meaningless.

File: stl/inc/ranges (hunk 16 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/ranges]: Even better.

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/ranges (hunk 11 above)
Comment (CaseyCarter) [reply stl/inc/ranges:4912]: We don't replace all uses of `forward` and `move` with `static_cast` due to readability concerns. As a special exception to that rule, we _do_ write `static_cast` instead of `forward` and `move` in concepts and constraints because compiler throughput concerns trump readability - at least until we could complete everything and get real numbers - and we want to avoid instantiating templates that may only ever be used in failing constraint checks.

Comment (CaseyCarter) [reply stl/inc/ranges:4912]: Coda: I didn't complain here - despite that this would seemingly violate the general guideline - since there are no competing overloads of "`operator()` with arity 2" and no `views_chunkable` concept. I wouldn't mind if you decide to change this for consistency, but I won't demand it either.

File: tests/std/tests/P2443R1_views_chunk_by/test.cpp [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <range_algorithm_support.hpp>
+using namespace std;
+
+constexpr auto equal_ranges = [](auto&& left, auto&& right) { return ranges::equal(left, right); };
+constexpr std::less<> pred{};
+using Pred = std::less<>;
```
Comment (StephanTLavavej) [inline tests/std/tests/P2443R1_views_chunk_by/test.cpp:19]: This test is already `using namespace std;`. Additionally, previously `Pred` was determined with `decltype` so it had to appear second. Now we can reverse the order to centralize the type.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I've pushed a conflict-free merge with `main` (as there had been a submodule update) and a trivial test change.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this chunky chocolate chip cookie! :cookie: :chocolate_bar: :smile_cat:
